src/Adapters/Storage/Postgres/PostgresStorageAdapter.js

```diff
@@ -1771,28 +1808,134 @@ export class PostgresStorageAdapter implements StorageAdapter {
-    return (conn || this._client).none('CREATE INDEX $1:name ON $2:name ($3:name)', [fieldName, className, type]);
+    conn = conn || this._client;
+    const self = this;
+    return conn.tx('create-indexes-if-needed', function * (t) {
+      if (type && (type.type === 'Polygon')) {
+        const index = {
+          [fieldName]: '2dsphere'
+        };
+
+        yield self.createIndex(className, index, 'gist', t);
+      }
+    });
+  }
+
+  createTextIndexesIfNeeded(className: string, query: QueryType, schema: SchemaType): Promise<any> {
+    for(const fieldName in query) {
+      if (!query[fieldName] || !query[fieldName].$text) {
+        continue;
+      }
+      let promise = Promise.resolve();
+      let existingIndexes = schema.indexes;
+      if (!existingIndexes) {
+        promise = this.setIndexesFromDB(className).then(() => {
+          return this.getClass(className);
+        });
+      }
+      return promise.then((dbSchema) => {
+        if (dbSchema) {
+          existingIndexes = dbSchema.indexes;
+        }
+        for (const key in existingIndexes) {
+          const index = existingIndexes[key];
+          if (index.hasOwnProperty(fieldName)) {
+            return Promise.resolve();
+          }
+        }
+        const indexName = `${fieldName}_text`;
+        const textIndex = {
+          [indexName]: { [fieldName]: 'text' }
+        };
+        return this.setIndexesWithSchemaFormat(className, textIndex, existingIndexes, schema.fields);
+      });
+    }
+    return Promise.resolve();
   }
 
   dropIndexes(className: string, indexes: any, conn: any): Promise<void> {
     const queries = indexes.map(i => ({query: 'DROP INDEX $1:name', values: i}));
     return (conn || this._client).tx(t => t.none(this._pgp.helpers.concat(queries)));
   }
 
-  getIndexes(className: string) {
+  dropAllIndexes(className: string) {
+    return this.getClass(className).then((schema) => {
+      const batch = [];
+      return this._client.tx(t => {
+        for (const key in schema.indexes) {
+          batch.push(t.none('DROP INDEX $1:name', key));
+        }
+        return t.batch(batch);
+      });
+    });
+  }
+
+  getIndexes(className: string, conn: ?any): Promise<any> {
     const qs = 'SELECT * FROM pg_indexes WHERE tablename = ${className}';
-    return this._client.any(qs, {className});
+    return (conn || this._client).any(qs, {className}).then((indexes) => {
+      return indexes.reduce((obj, index) => {
+        // Get field from index definition
+        const indexdef = index.indexdef.replace(/[\"]/g, "");
+        const regExp = /\(([^)]+)\)/;
+        let name = index.indexname;
+        const value = regExp.exec(indexdef)[1];
+        const fields = value.split(', ');
+        let key = {};
+        fields.forEach((field) => {
+          if (field === 'objectId') {
+            key = { _id: 1 };
+            name = '_id_';
```

Comment:
Doesn't this preclude compound indices that include `_id` as a field?

Reply:
I don't think so, I'll write a test case for it. Regarding https://github.com/parse-community/parse-server/issues/4719#issuecomment-386034171 can you show me which line you are talking about.

Reply:
In your [tests](https://github.com/parse-community/parse-server/pull/4629/files#diff-fe189dd02120c506a4daae7c8ddddb17R644) the schema you are expecting uses the `_id` field. While that is what backs objectId in the mongo backend, what I would expect Parse to store and report to the user is `_id_: { objectId: 1 }` since "objectId" is the name of the field in the parse interface. It makes it easier to write [tooling](https://github.com/HustleInc/parseconfig/) if the name in indexes matches the name in fields.

Reply:
It might introduce breaking change. I'm all for changing it tho.

Reply:
It is definitely a breaking change but it seems unlikely anybody is using index support right now since it basically doesn't work. Parse isn't even consistent in whether it returns the `_id_` index. @TylerBrock Thoughts?

Reply:
Index support currently works for Mongo, this PR is more geared towards Postgres so there will be some inconsistencies. 

If we are introducing breaking changes, I'd recommend removing the ability for users to add names to their indexes, and have parse autogenerate them. This makes it easier to keep track of duplicate indexes, transforming keys, and provide support for future DB adapters. Thoughts?

Reply:
It technically works but you can't create indices on most of the fields you would want to do so (pointers, updatedAt, etc).